pennylane/templates/tensornetworks/ttn.py

```diff
@@ -0,0 +1,210 @@
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+import warnings
+import pennylane as qml
+import pennylane.numpy as np
+from pennylane.operation import Operation, AnyWires
+
+
+def compute_indices(wires, n_block_wires):
+    """Generate a list containing the wires for each block.
+
+    Args:
+        wires (Iterable): wires that the template acts on
+        n_block_wires (int): number of wires per block
+    Returns:
+        layers (array): array of wire indices or wire labels for each block
+    """
+
+    n_wires = len(wires)
+
+    if n_block_wires % 2 != 0:
+        raise ValueError(f"n_block_wires must be an even integer; got {n_block_wires}")
+
+    if n_block_wires < 2:
+        raise ValueError(
+            f"number of wires in each block must be larger than or equal to 2; got n_block_wires = {n_block_wires}"
+        )
+
+    if n_block_wires > n_wires:
+        raise ValueError(
+            f"n_block_wires must be smaller than or equal to the number of wires; "
+            f"got n_block_wires = {n_block_wires} and number of wires = {n_wires}"
+        )
+
+    if not np.log2(n_wires / n_block_wires).is_integer():
+        warnings.warn(
+            f"The number of wires should be n_block_wires times 2^n; got n_wires/n_block_wires = {n_wires/n_block_wires}"
+        )
+
+    n_wires = 2 ** (int(np.log2(len(wires) / n_block_wires))) * n_block_wires
+    n_layers = int(np.log2(n_wires // n_block_wires)) + 1
+
+    layers = np.array(
+        [
+            [
+                wires[i]
+                for i in range(
+                    x + 2 ** (j - 1) * n_block_wires // 2 - n_block_wires // 2,
+                    x + n_block_wires // 2 + 2 ** (j - 1) * n_block_wires // 2 - n_block_wires // 2,
+                )
+            ]
+            + [
+                wires[i]
+                for i in range(
+                    x
+                    + 2 ** (j - 1) * n_block_wires // 2
+                    + 2 ** (j - 1) * n_block_wires // 2
+                    - n_block_wires // 2,
+                    x
+                    + 2 ** (j - 1) * n_block_wires // 2
+                    + n_block_wires // 2
+                    + 2 ** (j - 1) * n_block_wires // 2
+                    - n_block_wires // 2,
+                )
+            ]
+            for j in range(1, n_layers + 1)
+            for x in range(0, n_wires - n_block_wires // 2, 2 ** (j - 1) * n_block_wires)
+        ]
+    )
+
+    return layers
+
+
+class TTN(Operation):
+    """The TTN template broadcasts an input circuit across many wires following the architecture of a tree tensor network.
+    The result is similar to the architecture in `arXiv:1803.11537 <https://arxiv.org/abs/1803.11537>`_.
+
+    The argument ``block`` is a user-defined quantum circuit.``block`` should have two arguments: ``weights`` and ``wires``.
+    For clarity, it is recommended to use a one-dimensional list or array for the block weights.
```

Comment:
We have used fixed signatures a few times before in PennyLane, and always regretted it :) 

I don't know a better option, but one could at least allow `weights` (or better `parameters`) to be a list of inputs, which gets unpacked as `block(*parameters, wires=...)`. That's how `templates.broadcast` does it...

But it would be great to have a better solution one day!

Reply:
Thanks Maria! I made this change. I'm happy that it makes `block` more flexible.
I'm now getting some warnings about ragged arrays. I think somewhere deeper in PennyLane the parameters are converted to numpy arrays :thinking:

Reply:
That is very strange, because we should only ever work with lists of tensors/arrays, or with flattened versions. I think it would be good to trace where this error arises, just in case it is something trivial. Ideally one should never create a ragged array, they will be increasingly discontinued...

Reply:
I found that this happens with `qml.broadcast()` as well:
```
import pennylane as qml

def mytemplate(par1, par2, wires):
    qml.RY(par1[0], wires=wires[0])
    qml.RY(par1[1], wires=wires[0])
    qml.RY(par2,wires=wires[1])

dev = qml.device('default.qubit', wires=4)

@qml.qnode(dev)
def circuit(pars):
    qml.broadcast(unitary=mytemplate, pattern="chain", wires=[0,1,2,3], parameters=pars)
    return qml.expval(qml.PauliZ(0))

print(qml.grad(circuit)([[[1.0,1.0],2.0],[[1.0,1.0],2.0],[[1.0,1.0],2.0]]))
```
Error: array is not broadcastable to correct shape

Reply:
I traced the ragged array warning to the `construct()` method of  `qnode.py` :

```
def construct(self, args, kwargs):
        """Call the quantum function with a tape context, ensuring the operations get queued."""

        if self.interface == "autograd":
            # HOTFIX: to maintain backwards compatibility existing PennyLane code and demos, here we treat
            # all inputs that do not explicitly specify `requires_grad=False`
            # as trainable. This should be removed at some point, forcing users
            # to specify `requires_grad=True` for trainable parameters.
            args = [
                qml.numpy.array(a, requires_grad=True) if not hasattr(a, "requires_grad") else a
                for a in args
            ]
...
```
Here args is a qml.numpy.array made from `args`. `args` only has one element,`a`. This `a` is the ragged array supplied by a user.

I think it may be okay to leave things like this and put a warning in the documentation about the use of ragged arrays to define the template parameters :thinking:

Reply:
@DSGuala luckily we are planning to remove that HOTFIX if statement within days/weeks :)

Reply:
@antalszava tagging you here for visibility